### ISSUE TITLE
JSON API attribute name conversion when using a custom application serializer

### ISSUE THREAD

**Issue by begedin** [danielspaniel/ember-data-factory-guy#116]
In our application, we're using ember data with JSON API on the client side, Rails API with Active Model Serializers on the server side.

ActiveModelSerializers, by default, give us attributes in underscored form, so we had to override our application serializer like so:

```JavaScript

import Ember from 'ember';
import DS from 'ember-data';

var underscore = Ember.String.underscore;

export default DS.JSONAPISerializer.extend({
  keyForAttribute: function(attr) {
    return underscore(attr); // needed to transform AMS's underscores
  },

  keyForRelationship: function(rawKey) {
    return underscore(rawKey); // needed to transform AMS's underscores
  }
});

```

However, when we define a factory, we have to use camelized properties:

```JavaScript
FactoryGuy.define('user', {
 default: {
   admin: false,
   canTrial: false,
   email: 'johndoe@gmail.com',
   fullName: 'John Doe',
   id: 1
 }
});
```

From debugging, I'm seeing that what factory guy does is the following:

1. Get the factory
2. Run it through the default JSON API serializer to get a **dasherized** form
3. Pass it on to the store as an adapter payload
4. The store then uses the actual application serializer (our custom one), to convert back to the client camelized form.
5. Since the adapter expects an underscored form as input, the multi-word properties simply get ignored.

We could modify our adapter to account for this behavior and simply support both underscored and dasherized form of attribute keys, but I'm thinking FactoryGuy should use the application serializer anyway. Otherwise, it would prevent us from customizing the way our attribute (or relationship) keys are named.

Am I missing something, or is this really what's going on here?

**Comment by danielspaniel** [comment]
@begedin, you are almost 100% correct and missing very little, and I was waiting for someone to bring up this issue, because I knew it was sitting there waiting for someone to pour hot syrup all over it. 
  Unless you mean in item 2 that the 'default JSON API serializer' I am using it factory guy's own serializer, in which case you are 100% correct. 
  What should be happening is that I should use the application's serializer as you are suggesting. 
  Is there any chance that you could fix this and submit a PR? 
 I am on vacation starting tomorrow with not much internet for 10 days, otherwise I would give it a go. 

 In the JSONAPIAttributeTransformer class 

```javascript
var JSONAPIAttributeTransformer = function (/*store*/) {
  var defaultTransformFn = ''.dasherize;
```
I kind of set myself up to override the transform function .. ( sort of ) ... so that might be one angle, 
Or there might be a way to use the default application serializer rather than the transformer.
Though at least I think the change should be easy to make with the current set up, since it
 is very easy to make your own transformer and swap it with the current one.

**Comment by begedin** [comment]
@begedin: I might be able to find some time during the weekend to fix this, but I can't make any promises. In any case, thanks for the pointer and the quick reply. If I managed to find time to make a PR before someone else does, I'll do it.

**Comment by danielspaniel** [comment]
Thanks @begedin for offering to fix, and I am pretty sure you can get it pretty quickly because of the way the JSONAPIAttributeTransformer is setup .. I think it will be fun to try and if it fails, oh .. well .. just wait for me to get back

**Comment by begedin** [comment]
I know my PR has been merged, but looking at the AMS attribute transformer, I'm thinking the same approach should be used there, since the underscore default is used in that one.

**Comment by danielspaniel** [comment]
that sounds correct to me @begedin .. thanks for reminding me about that ( I thought about it but forgot ) .. I will change that one too

**Comment by danielspaniel** [comment]
ok .. this is now fixed in v1.13.8

**Comment by danielspaniel** [comment]
But let me know if you see other issues
